File: server/proto/actions.go
package proto

type ActionGPIO struct {
	Pin int `json:"pin"`
	// "high", "low" or "pulse".
	Mode       string `json:"mode"`
	DurationMs int    `json:"durationMs"`
	// For relay boards, which usually activate when pulled low.
	ActiveLow bool `json:"activeLow"`
	// Non-empty when the firmware already drives this pin. Advisory only.
	ReservedFor string `json:"reservedFor,omitempty"`
}

type ActionInfo struct {
	Id   string `json:"id"`
	Name string `json:"name"`
	// "gpio" or "command".
	Type       string      `json:"type"`
	GPIO       *ActionGPIO `json:"gpio,omitempty"`
	Command    string      `json:"command,omitempty"`
	ShowInMenu bool        `json:"showInMenu"`
}

type GetActionsRsp struct {
	Actions    []ActionInfo `json:"actions"`
	ShortPress string       `json:"shortPress"`
	LongPress  string       `json:"longPress"`
	// The daemon's own button handling continues alongside a custom action.
	KeepDefaults bool `json:"keepDefaults"`
	// The usable gpio range on this hardware, so the UI can bound its input.
	MinGPIO int `json:"minGpio"`
	MaxGPIO int `json:"maxGpio"`
}

type SetActionsReq struct {
	Actions      []ActionInfo `json:"actions"`
	ShortPress   string       `json:"shortPress"`
	LongPress    string       `json:"longPress"`
	KeepDefaults bool         `json:"keepDefaults"`
}

type RunActionReq struct {
	Id string `json:"id" form:"id" validate:"required"`
}

File: web/src/api/actions.ts
import { http } from '@/lib/http.ts';

export type ActionGPIO = {
  pin: number;
  // 'high', 'low' or 'pulse'
  mode: string;
  durationMs: number;
  // Relay boards usually activate when the pin is pulled low.
  activeLow: boolean;
  // Set when the firmware already drives this pin. Advisory only.
  reservedFor?: string;
};

export type ActionInfo = {
  id: string;
  name: string;
  // 'gpio' or 'command'
  type: string;
  gpio?: ActionGPIO;
  command?: string;
  showInMenu: boolean;
};

export type ActionsConfig = {
  actions: ActionInfo[];
  shortPress: string;
  longPress: string;
  keepDefaults: boolean;
  minGpio: number;
  maxGpio: number;
};

export function getActions() {
  return http.get('/api/actions');
}

export function setActions(config: {
  actions: ActionInfo[];
  shortPress: string;
  longPress: string;
  keepDefaults: boolean;
}) {
  return http.post('/api/actions', config);
}

export function runAction(id: string) {
  return http.post('/api/actions/run', { id });
}
